src/main/java/com/es/phoneshop/controller/servlet/ProductDetailsPageServlet.java

```diff
@@ -1,31 +1,94 @@
 
 public class ProductDetailsPageServlet extends HttpServlet {
-    private static final String PRODUCT_CODE_ATTRIBUTE = "product";
-    private static final String ERROR_MESSAGE_ATTRIBUTE = "message";
+    private static final String ERROR_NOT_FOUND = "Product with code '%s' not found.";
+    private static final String ERROR_NOT_ENOUGH_STOCK = "Not enough stock. Available:%s";
+    private static final String ERROR_NOT_NUMBER = "Not a number";
+    private static final String REQUEST_ATTRIBUTE_PRODUCT = "product";
+    private static final String REQUEST_ATTRIBUTE_MESSAGE = "message";
+    private static final String REQUEST_ATTRIBUTE_ERROR = "error";
+    private static final String REQUEST_ATTRIBUTE_ADDING_TO_CART_MESSAGE = "show";
+    private static final String REQUEST_PARAM_QUANTITY = "quantity";
+    private static final String REDIRECT_AFTER_ADDING_TO_CART = "%s/products/%s?message=Added to cart successfully";
     private static final String PRODUCT_DETAILS_PATH = "/WEB-INF/pages/productDetails.jsp";
-    private static final String ERROR_MESSAGE_PART = "Product with code '%s' not found.";
     private static final int NOT_FOUND_ERROR_CODE = 404;
-    private ArrayListProductDao arrayListProductDao = ArrayListProductDao.getInstance();
+    private ArrayListProductDao dao = ArrayListProductDao.getInstance();
+    private CartService cartService = CartService.getInstance();
+    private RecentlyViewedService recentlyViewedService = RecentlyViewedService.getInstance();
 
     @Override
-    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
-        String productId = request.getPathInfo().substring(1);
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        Optional<Product> product = extractProductOrSendError(req, resp);
+        if (product.isEmpty()) {
+            return;
+        }
+        req.setAttribute(REQUEST_ATTRIBUTE_PRODUCT, product.get());
+        req.getRequestDispatcher(PRODUCT_DETAILS_PATH).forward(req, resp);
+        recentlyViewedService.updateRecentlyViewedLine(req.getSession(), product.get());
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        Optional<Product> product = extractProductOrSendError(req, resp);
+        Optional<Integer> quantity = extractQuantityOrForward(req, resp);
+        if (product.isEmpty() || quantity.isEmpty()) {
+            return;
+        }
+        addProductToCartOrForward(req, resp, product.get(), quantity.get());
+    }
+
+    private Optional<Product> extractProductOrSendError(HttpServletRequest req, HttpServletResponse resp)
+            throws IOException {
+        String productId = req.getPathInfo().substring(1);
         try {
-            request.setAttribute(PRODUCT_CODE_ATTRIBUTE, arrayListProductDao.findProduct(Long.valueOf(productId)));
```

Comment:
You might put here as well
request.setAttribute(REQUEST_ATTRIBUTE_PRODUCT, ...)
and rename the method to smth like (attachProductOrSendError)

Reply:
In this case, you won't need to invoke doGet from extractQuantityOrForward & addProductToCartOrForward. You'll just need to do request.getRequestDispatcher(..).forward() instead